gh-127350: Add Py_fopen() function (#127821)
python/cpython: Doc/c-api/sys.rst

      The function now uses the UTF-8 encoding on Windows if
      :c:member:`PyPreConfig.legacy_windows_fs_encoding` is zero.

.. c:function:: FILE* Py_fopen(PyObject *path, const char *mode)

>    Similar to the :c:func:`!fopen` function, but *path* is a Python object and

Reviewer: Another nit; you can consider simplifying this.

```suggestion
   Similar to :c:func:`!fopen`, but *path* is a Python object and
```

   an exception is set on error.

>    *path* must be a :class:`str` object or a :class:`bytes` object.

Reviewer 2: It can also be a path-like object, isn't? But its support is inconsistent on Windows.

Author: Fixed.

>    On success, return the new file object.

Reviewer 2: It is not an object. This function returns a FILE pointer.

   On error, set an exception and return ``NULL``.

   The file must be closed by :c:func:`Py_fclose` rather than calling directly

>    ``fclose()``.

Reviewer: Nit.

```suggestion
   :c:func:`!fclose`.
```

   The file descriptor is created non-inheritable (:pep:`446`).

   The caller must hold the GIL.

   .. versionadded:: next

.. c:function:: int Py_fclose(FILE *file)

>    Closes files that were opened by :c:func:`Py_fopen`.

Reviewer: Imperative mood.

```suggestion
   Close a file that was opened by :c:func:`Py_fopen`.
```

>    Call ``fclose(file)``.

Reviewer 3: I know it's mentioned in the `Py_fopen` docs, but for people who just read this one, it's worth adding a bit more clarity. (Probably should also explain the return value... it's just `errno` or 0, right?)

```suggestion
   Closes files that were opened by :c:func:`Py_fopen`.
```

Author: > (Probably should also explain the return value... it's just errno or 0, right?)

Done.
